### ISSUE TITLE
There is currently no way to specify features automatically when using `cargo doc`

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/cargo#8905]
I thought at first about using `[profile.doc]`, but it is doing nothing (as displayed by cargo itself) and there are no other way to do so.

Then I looked a bit deeper and noticed that you can't set features under profiles, making this not the right approach.

Therefore, I suggest adding a new section simply called "rustdoc" which would allow for now:

```toml
[rustdoc]
features = ["whatever"]
```

**Comment by Andlon** [comment]
In an attempt to motivate this issue, I'd like to bring up the example of proc macros that add or modify documentation, of the form
```rust
/// Foobaring a Baz.
#[modify_docs]
fn foobar(baz: Baz) {}
```
or
```rust
//! Crate docs
#![doc = generate_super_nice_html!()]
```

Or, for a more realistic example, consider the [embed-doc-image](https://crates.io/crates/embed-doc-image) crate:
```rust
//! A collection of crustaceans.
//!
//! # Ferris
//! ![Ferris](ferris)
//!
//! Ferris is a crab, as you can see in the above image.
#![doc = embed_image!("ferris", "images/ferris.png")]
```
A bit of a problem with these solutions is that - although the macros shown in the above examples perhaps *only* modify documentation, they still need to run for normal compilation. This means that:
-  they might need to pull in dependencies (not dev-dependencies) that are not needed by the code in a non-documentation setting
- they generally still need to run in a non-documentation setting

For example, `embed_image!` will load an image from disk and encode it as `base64`, which might be expensive depending on the image and hardware, and as such it's conceivable that it might impact compile times in non-documentation code.

There is no *truly* good workaround for these kind of macros at the moment. Sure, you can define a feature that you need to manually enable when calling `rustdoc`, but this:
- is really annoying.
- the feature might be needed by a dependency and not something you can change from the top-level crate. So this will generally tend to break documentation for dependencies.

**Comment by joshtriplett** [comment]
We talked about this at length in today's @rust-lang/cargo meeting. Copying some text from #10435 to move the discussion here:

Currently, Rust has three kinds of dependencies: dependencies, dev-dependencies, and build-dependencies. Adding a fourth kind of dependency would be a *substantial* increase in complexity.

The original issue, https://github.com/rust-lang/cargo/issues/8905 , started out talking about using feature flags for this, and in our discussions today we felt like that approach would be much simpler and make more use of existing Cargo facilities.

Crates today already use features for this, and then add docs.rs metadata telling docs.rs to use those features. And features can already control dependencies or features of dependencies.

We're open to a new mechanism to automatically enable a feature flag on doc builds and doctests, so that `cargo doc` and `cargo test` work automatically. We'd be happy to discuss different ways to make that work.

Also, separately, this is something we'd consider to be a [large feature](https://doc.crates.io/contrib/process/index.html#working-on-large-features), for which we'd like to see an RFC design rather than going directly to a PR. Doing design iteration in a PR is much higher bandwidth for us than reviewing an RFC.

I'm going to close this PR in favor of https://github.com/rust-lang/cargo/issues/8905 , and propose that we discuss the feature design further there and subsequently in an RFC.

**Comment by Andlon** [comment]
It seems to me that this issue stems from the fact that documentation and standard builds have different sets of desirable *default* feature sets. For example, a crate might only want to use a minimal set of features for a dependency, but when building documentation for a crate higher in the dependency tree, we still want to have the "proper" documentation for the dependencies.

However, currently there is only a single mechanism for specifying default features, with no way to distinguish between standard and doc builds.

With that in mind, it seems at first glance to me that something like the following would be ideal:

```toml
[features]
default = ["my", "default", "features"]

# Document all the default stuff, but also make sure that we
# get serde docs
doc-default = [ "default", "serde" ]
```
I think `doc-default` should ignore `--no-default-features` (command-line) and `default-features = false` (Cargo.toml). Otherwise documentation of dependencies might be incomplete when not using default features in the build.

Instead, we would need a separate `doc-default-features = false` flag (and command-line argument) so that dependencies can fine tune what they want to document from dependencies.

Certainly the latter is a breaking change: having doc builds treat `default-features = false` differently is a breaking change in cargo's behavior when building documentation. I'm not able to see the complete ramifications of this at the moment, and it would be great if perhaps someone from the @rust-lang/cargo team could comment on that.

**Comment by GuillaumeGomez** [comment]
Having `doc-default` would also allow to handle doc dependencies pretty easily. I like this suggestion!

**Comment by joshtriplett** [comment]
What is the advantage of having a doc-default feature, rather than just enabling dev-dependencies for all doc builds? doctests already have to install dev-dependencies, so this would just enable those for docs as well.

**Comment by joshtriplett** [comment]
You do need it to build your docs, which you have to do to as part of `cargo test` to run doctests. And dev-dependencies typically includes many heavier-weight things that are left out of ordinary dependencies; for instance, an async crate might pull in a full async runtime and all its dependencies for testing.

As long as they're not pulled in for ordinary builds, I don't think we need to worry much about the weight of dev-dependencies.

**Comment by Andlon** [comment]
> You do need it to build your docs, which you have to do to as part of `cargo test` to run doctests. And dev-dependencies typically includes many heavier-weight things that are left out of ordinary dependencies; for instance, an async crate might pull in a full async runtime and all its dependencies for testing.
> 
> As long as they're not pulled in for ordinary builds, I don't think we need to worry much about the weight of dev-dependencies.

I don't immediately see how making doc builds solves the problem here. Consider my motivating example for [embed-doc-image](https://github.com/rust-lang/cargo/issues/8905#issuecomment-1041766664):

```rust
//! ![Ferris](ferris)
#![doc = embed_image!("ferris", "images/ferris.png")]
```
Here we'll need the `embed_image!` macro in all builds to make this compile at all. So that means that we can not simply make `embed-doc-image` a `dev-dependency`: it needs to be present in normal builds as well.

In addition to this, there may be a compile-time cost associated with invoking the macros as well. For example, with `embed-doc-image` you would need, for each image you use in your documentation, to:

- load a file (possibly "large") from the file system to memory
- encode the image as `base64`

So even if this would only happen in `test` settings, it could *perhaps* add a sizeable cost to the debug-compile-run cycle (depending on how well incremental compilation does its job).

**Comment by Andlon** [comment]
@Urgau: I agree that would be almost ideal *if* it worked as you would expect (see my [above comment](https://github.com/rust-lang/cargo/issues/8905#issuecomment-1056827336), which I slipped in just before you posted yours!).

Your latter suggestion is interesting... I guess someone with knowledge of how this propagates through cargo/rustdoc would have to chime in to say if that is workable.

**Comment by Andlon** [comment]
> > You're right, that's weird. Is this a feature ? because I would consider this as a bug.
> 
> This is a bug and the main motivation behind this issue. :)

Sorry, but I believe this statement might confound this issue somewhat: even if `cfg(doc)` worked the way you might expect (propagates to dependencies during doc compilation etc.), it still wouldn't allow us to enable certain features or dependencies only for doc builds.

**Comment by GuillaumeGomez** [comment]
You are right on that. This is why I was hoping for `doc-default` to kinda fix both. But yes, currently we can't specify with cargo doc dependencies nor features.

**Comment by GuillaumeGomez** [comment]
Yes, it's as expected and why I think @joshtriplett is suggesting to re-use `dev-dependencies`.

**Comment by Andlon** [comment]
I will also add that I think if `dev-dependencies` + `cfg(doc)` actually work out, then I believe this might be better than to add additional complexity with e.g. the `doc-default` suggestion. I think we should really not encourage crates to further complicate builds with complex behavior when building docs.

**Comment by GuillaumeGomez** [comment]
It's not always an option unfortunately. For example, in `gtk-rs` we have a `dox` feature which updates the source code to add the (LGPL) documentation. If we don't have a way to enable this feature only on doc build, then it means your code will be updated every time you run `cargo test`. Or what we did: we now generate empty documentation on docs.rs and host the documentation ourselves.

**Comment by Andlon** [comment]
@GuillaumeGomez: forgive me if I'm being too naive here, but wouldn't `cfg(doc)` solve that problem?

**Comment by Urgau** [comment]
Okay, so if I take everything in account I propose adding:
```toml
[profile.doc]
features = ["feature", "automaticly", "activated", "for", "doc"]
```
With `[dev-dependencies]` also available when documenting.
And with `--cfg doc` being applied for every dependencies (normal or dev) when documenting.

**Comment by GuillaumeGomez** [comment]
> @GuillaumeGomez: forgive me if I'm being too naive here, but wouldn't `cfg(doc)` solve that problem?

Sorry but I'm the one who's lost: how `cfg(doc)` would fix the re-compilation issue if the source code is updated?

**Comment by Andlon** [comment]
@GuillaumeGomez: I think I didn't understand what you meant. But couldn't `cargo doc --features dox` be replaced with `cfg(doc)` directives? For example to do the codegen in a build script when `cfg(doc)`? Or is `cfg(doc)` set when running `cargo test` as well?

**Comment by GuillaumeGomez** [comment]
No because we need to do things in `build.rs` (downloading docs and putting it back into the source code). But I think it's not an issue. The big issue in gtk-rs currently is that we handle versions with features and we need dependencies to have been built with them and with the `dox` feature to make it work. So if they are built with `cfg(doc)`, almost all of our problems are solved.

Being able to have doc specific features would be a big plus too.

> Or is `cfg(doc)` set when running `cargo test` as well?

It is when you're running `cargo test --doc` with `cfg(doctest)` too. But it's only for the current crate, not its dependencies.

**Comment by Manishearth** [comment]
> What is the advantage of having a doc-default feature, rather than just enabling dev-dependencies for all doc builds? doctests already have to install dev-dependencies, so this would just enable those for docs as well.

Because enabling dev-deps for all dependencies will probably pull in all kinds of stuff that in aggregate is likely to have build issues. `dev-dependencies` is currently quite squarely in the zone of "users of this crate don't need to care about this", and this changes that.

**Comment by joshtriplett** [comment]
> `cfg(doc)` _would_ seem like the ideal solution for the above problem, except for the fact that it doesn't play well with dependencies: roughly speaking, `cfg(doc)` only applies to the top-level crate but not for the documentation generated for dependencies.

That...seems like a bug? Unless I'm misunderstanding how `cfg(doc)` is supposed to work.

**Comment by Urgau** [comment]
The problem with `cfg(doc)` not being applied is because when a crate has a dependency this dependency first needs to be check (cargo term) then passed to `rustdoc` BUT when checking cargo doesn't add `--cfg doc`.

**Comment by Nemo157** [comment]
Changing `cfg(doc)` to be passed down to all dependencies is very likely to be a breaking change. I know `async-std` in particular is doing some very cursed things based on `cfg(doc)` so I would not be surprised if it broke doc builds of its dependents. A quick test of `RUSTFLAGS=--cfg=doc cargo check` actually reveals an earlier failure when checking `socket2`:

```
error[E0599]: no method named `_set_nosigpipe` found for reference `&socket::Socket` in the current scope
    --> /home/nemo157/.cargo/registry/src/github.com-1ecc6299db9ec823/socket2-0.4.4/src/sys/unix.rs:1123:14
```

Maybe it would be possible to tie this to an edition change or some other opt-in to the breakage.

One other big consideration is that `rustdoc` is not the only consumer of the `doc` attributes. These are also consumed by other development tools [such as rust-analyzer](https://rust-analyzer.github.io/manual.html#hover). For those tools it's actually better if during development all builds get the "complete docs", not just when the crate is built with `cfg(doc)` or `dev-dependencies`.

**Comment by Nemo157** [comment]
> roughly speaking, `cfg(doc)` only applies to the top-level crate but not for the documentation generated for dependencies.

To make this a little more precise, `cfg(doc)` applies only during the `rustdoc` build for a crate, so when you `cargo doc` and it documents all crates in your dependency tree they will each have `--cfg=doc` implied while their docs are built; _but_ when your crates documentation is built it uses the standard `rmeta` from each of its dependencies that was built without `--cfg=doc`.

You can see the difference when inlining documentation, if you inline documentation from a dependency that does something like `#[cfg_attr(doc, doc = "foo")]` then `foo` will appear in the docs generated for their crate but not yours. (Based on a quick grep of all crates I have locally, this is not something that happens, but it is likely to be one outcome of allowing the sort of features that are wanted here).

**Comment by Nemo157** [comment]
There was also some prior discussion of a different usecase of something like `doc-dependencies` in <https://github.com/rust-lang/rust/issues/74481>.
